test/basic_json.cpp

```diff
@@ -200,7 +200,8 @@ TEST(general, json_stream_operator)
     t.rootref() << m;
     std::stringstream ss;
     ss << as_json(t);
-    Tree res = c4::yml::parse(to_csubstr(ss.str()));
+    std::string result = ss.str();  // keep the memory in scope!
+    Tree res = c4::yml::parse(to_csubstr(result));
     m.clear();
     res.rootref() >> m;
     EXPECT_EQ(m["foo"], 1);
```

Comment:
Ha! This here is the problem... You are testing the original map: `m` is the map. You need to test the resulting tree...
So for instance, this
```c++
EXPECT_EQ(m["foo"], 1);
```
should be instead
```c++
EXPECT_EQ(res["foo"].val(), "1");  // note that "1" is and must be a string
```

This also explains why you were not seeing a crash (RAF, read after free) even after the string is no longer small, and hence using heap memory. That memory was not read from again **after** the parse() statement. Inside the parse() statement the temporary is still valid. As soon as the parse() returns, the temp is destroyed. If then you had read from it in the EXPECTs, you would have a crash.

You can also cause the RAF crash to occur inside parse if you do the following:
```c++
c4::csubstr s = ss.str();
// here, the temporary ss.str() was already destroyed so is pointing at stale memory
Tree res = c4::yml::parse(s);  // crash, RAF
```

Reply:
Well, I was wrong, sorry. Your line
```c++
res.rootref() >> m;
```
Does read from the memory. And indded, in my VS solution the test succeeds.

Reply:
Also tried with g++7 and clang10, valgrind and all the sanitizers, and it also succeeds. Weird.

I was using acquired wisdom, eg [this](https://stackoverflow.com/questions/4214153/lifetime-of-temporaries) or [this](https://stackoverflow.com/questions/4267676/do-temporary-objects-have-scope).

Reply:
Oh... I know.

Your call is `parse(to_csubstr(ss.str()));`. The `ss.str()` call creates a temporary. Indeed, the scope of this temporary is that of this expression, all the way to the `;`. So just like I said above, the temporary string will get destroyed AFTER parse() returns. Inside this expression, your call to `to_csubstr()` creates an _immutable_ string view. This view is valid inside the call to parse(). So parse will successfully finish.

But here is the detail that matters: parse() is given an immutable string. Note the ryml parser cannot parse directly from an immutable string. Because of YAML block scalars, block indentation and quote escaping behavior, ryml needs to modify the YAML source as it reads. So, when it is given the immutable string, ryml first copies into the tree's arena and parses from there. Then, all subsequent reads are viewing the tree's arena instead of the string's memory.

So that code is safe, because it is reading from memory in the tree's arena!

However.

If you slightly change this to use `to_substr(ss.str())` instead of `to_csubstr(ss.str())`, then you will indeed get the RAF from memory dangling out of the destroyed string. That's because now you've passed the mutable string view, which ryml will read from directly. So the parse will still succeed, but then as soon as you read from the tree via the `>>` operator, it will RAF.

So the recommendation still stands: when creating views to other objects, be careful with those other objects' lifetimes with the respect to the view object. Give names to those objects, and make sure their lifetime supersedes the view's lifetime. Views are extremely efficient and empowering, but they do have this mental overhead with the lifetime responsibility.

In this case, give a name to the temporary, and all the risk goes away, without any cost other than the extended lifetime of the temporary.

There. Sorry for misreading some of your code's details, and in particular this important mutability detail.

I'll have to review the API to make it clear to the caller that the immutable csubstr is first copied to the tree's arena. I fully admit that this situation is unsatisfactory.

Reply:
FYI, this is the test I adapted from your code:
```c++
TEST(general, json_stream_operator)
{
    std::map<std::string, int> out, m({{"bar", 2}, {"foo", 1}, {"foobar_barfoo:barfoo_foobar", 1001}, {"asdfjkl;", 42}, {"00000000000000000000000000000000000000000000000000000000000000", 1}});
    Tree t;
    t.rootref() << m;
    std::stringstream ss;
    ss << as_json(t);
    Tree res = c4::yml::parse(to_substr(ss.str()));
    EXPECT_EQ(res["foo"].val(), "1");
    EXPECT_EQ(res["bar"].val(), "2");
    EXPECT_EQ(res["foobar_barfoo:barfoo_foobar"].val(), "1001");
    EXPECT_EQ(res["asdfjkl;"].val(), "42");
    EXPECT_EQ(res["00000000000000000000000000000000000000000000000000000000000000"].val(), "1");
    res.rootref() >> out;
    EXPECT_EQ(out["foo"], 1);
    EXPECT_EQ(out["bar"], 2);
    EXPECT_EQ(out["foobar_barfoo:barfoo_foobar"], 1001);
    EXPECT_EQ(out["asdfjkl;"], 42);
    EXPECT_EQ(out["00000000000000000000000000000000000000000000000000000000000000"], 1);
}
```
Like I said, this fails spectacularly because of the RAF. To make it succeed, just do
```c++
TEST(general, json_stream_operator)
{
    std::map<std::string, int> out, m({{"bar", 2}, {"foo", 1}, {"foobar_barfoo:barfoo_foobar", 1001}, {"asdfjkl;", 42}, {"00000000000000000000000000000000000000000000000000000000000000", 1}});
    Tree t;
    t.rootref() << m;
    std::string str;
    {
        std::stringstream ss;
        ss << as_json(t);
        str = ss.str();
    }
    Tree res = c4::yml::parse(to_substr(str));
    EXPECT_EQ(res["foo"].val(), "1");
    EXPECT_EQ(res["bar"].val(), "2");
    EXPECT_EQ(res["foobar_barfoo:barfoo_foobar"].val(), "1001");
    EXPECT_EQ(res["asdfjkl;"].val(), "42");
    EXPECT_EQ(res["00000000000000000000000000000000000000000000000000000000000000"].val(), "1");
    res.rootref() >> out;
    EXPECT_EQ(out["foo"], 1);
    EXPECT_EQ(out["bar"], 2);
    EXPECT_EQ(out["foobar_barfoo:barfoo_foobar"], 1001);
    EXPECT_EQ(out["asdfjkl;"], 42);
    EXPECT_EQ(out["00000000000000000000000000000000000000000000000000000000000000"], 1);
}
```